Feature: replace RnDiffApp with entered AppName

Naturalclar opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please follow the template so that the reviewers can easily understand what the code changes affect -->

# Summary

This PR is part of #108 
This adds a text input in the header which allows user to enter their AppName.
All the `RnDiffApp` and `rndiffapp` will be replaced with the entered AppName (in lowercase if needed)

<!--
Explain the **motivation** for making this change: here are some points to help you:

* What issues does the pull request solve? Please tag them so that they will get automatically closed once the PR is merged
* What is the feature? (if applicable)
* How did you implement the solution?
* What areas of the website does it impact?
-->

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

![Kapture 2019-12-27 at 20 39 44](https://user-images.githubusercontent.com/6936373/71516128-9fff7a00-28ea-11ea-88dc-66fba4e29a53.gif)


## What are the steps to reproduce?

## Checklist

<!-- Check completed item, when applicable, via: [X] -->

- [x] I tested this thoroughly

pvinis commented:
so good! awesome. I'll test it when I'm home before merging.

lucasbento commented:
Hi @Naturalclar, thank you for opening this PR.

The functionality is pretty useful, however, I would rather have the input for the application name to be under a button (the cog on the top right, for instance) as this is not required for everyone and can be just noise for most of the users.

About the warning on the bottom, I like it, I would suggest having it:

- Only showing when the user clicks on the `Show me how to upgrade!` button;
- As an info alert, [you can check it here](https://ant.design/components/alert/);
- With a button to close it which will only show it again if the user reloads/reopens the page (using state).

What do you think?

Naturalclar commented:
@lucasbento @pvinis 
Thanks for the feedback!
I've placed the AppName input inside the setting PopOver.
What do you think?

![Kapture 2019-12-29 at 22 55 01](https://user-images.githubusercontent.com/6936373/71557789-7a12da80-2a8e-11ea-93b8-30b9b425999f.gif)

As for the warning on the bottom, it's not part of this PR, but I'll gladly make another PR that makes the change you request for :)

pvinis commented:
Pretty great! I have two things to comment on.

- I would like a placeholder for the input field, some like `MyAwesomeApp`, so users can see the usual capitalization.
- When I type fast in there, the whole diff view refreshes and some of the time the result is not correct, like
<img width="1223" alt="Screen Shot 2019-12-30 at 10 48 10" src="https://user-images.githubusercontent.com/100233/71574560-ed016d00-2af1-11ea-98a8-9a8c18533db3.png">

Naturalclar commented:
> I would like a placeholder for the input field, some like MyAwesomeApp, so users can see the usual capitalization.

Sounds good! I've added a placeholder to the input form.

<img width="346" alt="Screen Shot 2019-12-31 at 2 23 42 PM" src="https://user-images.githubusercontent.com/6936373/71611109-f2d27d80-2bd9-11ea-9825-2b228e6a860d.png">

>When I type fast in there, the whole diff view refreshes and some of the time the result is not correct

I'm wondering what would be the best way to make the form without this error happening.
Should I place a buton inside settings to set the AppName so that refetching won't happen during input?

lucasbento commented on src/components/common/DiffViewer.js:
```diff
@@ -61,6 +62,16 @@ const DiffViewer = ({
     localStorage.setItem('viewStyle', newViewStyle)
   }
 
+  const replaceAppName = useCallback(
```
What's the reason for using `useCallback`?

Naturalclar replied:
The current lint setup automatically adds `replaceAppName` as dependency of `useEffect` in DiffViewer.js
Since a function declaration is a different instance on every render, using `replaceAppName` without using `useCallback` to memoize the function will result in an infinite loop.

lucasbento commented on src/components/common/Settings.js:
```diff
@@ -14,11 +14,24 @@ const Settings = ({ handleSettingsChange }) => {
-            <Checkbox value={SHOW_LATEST_RCS}>{SHOW_LATEST_RCS}</Checkbox>
-          </div>
-        </Checkbox.Group>
+        <Form>
```
Why do you need to use `Form`?

Naturalclar replied:
I was looking for a good way to style the contents inside the settings, and Form seemed to have what the list-esque style that I was looking for.
It's not a requirement to use Form.

lucasbento commented on src/components/pages/Home.js:
```diff
@@ -61,6 +61,7 @@ const Home = () => {
   const [settings, setSettings] = useState({
     [`${SHOW_LATEST_RCS}`]: false
   })
+  const [appName, setAppName] = useState('')
```
Is there a specific reason to not use `null` as the initial state?

Naturalclar replied:
Personally, I prefer using empty string over `null`, as `useState` infers that `appName` will be a typeof `string`, and it's (imo) it's clear to the 
 reader that this will be a string.
It's also to keep code consistency as other places that uses `useState` for string seems to initiate the value with an empty string as well.
If I were to write this in typescript with `null`, I'd write it

```
const [appName, setAppName] = useState<string | null >(null)
```

and there'd have to be a null check all over the place, which makes the code harder to read (imo).

It's not too big of a deal and I'm fine with using `null` for initialization.

pvinis replied:
I also prefer null but I guess for now `''` is fine for the reasons you mentioned. Maybe part of the rework should be typescript too? 🤔 Then we could do the template thing you mentioned. I'll add it.

lucasbento replied:
> Personally, I prefer using empty string over null, as useState infers that appName will be a typeof string, and it's (imo) it's clear to the
reader that this will be a string.

I like this 🙂

Naturalclar commented:
@pvinis @lucasbento 
Thanks so much for the feedback!
I've adjusted the style of the input a bit, and fixed the problem for typing fast.
the setTimeout seem to have done the trick! So awesome!

![Kapture 2020-01-03 at 0 32 18](https://user-images.githubusercontent.com/6936373/71675291-c3ca3080-2dc0-11ea-80f2-043856d80071.gif)

kelset commented:
This looks so great ❤️ thanks @Naturalclar 👏
